Replace `reboot_command` integration test with unit tests

webknjaz opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
$sbj

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Test Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
test/integration/targets/reboot
test/units/plugins/action/test_reboot.py

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
Requested in https://github.com/ansible/ansible/pull/78402#pullrequestreview-1059403483.

webknjaz committed:
Add unit tests for the reboot action plugin

webknjaz committed:
🔥 Drop the `reboot_command` integration test

s-hertel commented:
@webknjaz Is this still something you're working on?

webknjaz commented:
@s-hertel it is complete but is awaiting @mattclay to make a judgment call on whether we want this merged at all.

mattclay commented on test/units/plugins/action/test_reboot.py:
```diff
@@ -0,0 +1,145 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
```
This boilerplate isn't needed for controller-only unit tests any more:

```suggestion
import os
```

mattclay commented on test/units/plugins/action/test_reboot.py:
```diff
@@ -0,0 +1,145 @@
+def test_reboot_command(action_plugin, mocker, monkeypatch):
+    """Check that the reboot command gets called and reboot verified."""
+    def _patched_low_level_execute_command(cmd, *args, **kwargs):
+        if cmd == _SENTINEL_TEST_COMMAND:
+            return {
+                'rc': 0,
+                'stderr': '<test command stub-stderr>',
+                'stdout': '<test command stub-stdout>',
+            }
+
+        if cmd == _SENTINEL_REBOOT_COMMAND:
+            return {
+                'rc': 0,
+                'stderr': '<reboot command stub-stderr>',
+                'stdout': '<reboot command stub-stdout>',
+            }
+
+        raise RuntimeError(cmd)  # This should never be hit in the test
```
Refactoring this avoids unreachable code and gets coverage of this file to 100%:

```suggestion
        return {
            _SENTINEL_TEST_COMMAND: {
                'rc': 0,
                'stderr': '<test command stub-stderr>',
                'stdout': '<test command stub-stdout>',
            },
            _SENTINEL_REBOOT_COMMAND: {
                'rc': 0,
                'stderr': '<reboot command stub-stderr>',
                'stdout': '<reboot command stub-stdout>',
            },
        }[cmd]
```

webknjaz replied:
@mattclay I kinda wanted this to be here as a precaution for future changes. Can we do some `# pragma: no cover` here instead?

mattclay replied:
@webknjaz What do you mean by "precaution for future changes"?  Both approaches result in an exception if `cmd` is unknown.

Use of `# pragma: nocover` should be limited to cases where there's no reasonable way to get coverage. There are a few cases where using it does make sense. A few examples (taken from our own tests):

* In `if t.TYPE_CHECKING` branches (which should also be used sparingly).
* In methods required to implement abstract types where the method raises `NotImplementedError`.

For most other uses, flow control changes (usually minor) can avoid the need for a pragma entirely.

webknjaz replied:
Agreed, you're right. With the example you posted, the exception turned from an explicit `RuntimeError` into an implicit `KeyError`, and that's what confused me initially since I'd lost some context. Honestly, I'd prefer an explicit `ThisShoudntBeHappening` but I agree that sticking pragmas and noqas everywhere is a bad idea and I tend to follow the same principle you described.

I'll adopt your suggestion.

mattclay commented on test/units/plugins/action/test_reboot.py:
```diff
@@ -0,0 +1,145 @@
+        _patched_low_level_execute_command,
+    )
+
+    action_plugin._low_level_execute_command = _patched_low_level_execute_command
```
This appears to be redundant with the `monkeypatch.setattr` above:

```suggestion
```

webknjaz replied:
This is weird... I must've left it here by accident. Let my try applying this change.

mattclay reviewed (dismissed):
It looks like removing the `reboot_command` integration tests will cut at least 80 seconds from the tests running on RHEL 9.1.

Overall the code coverage is fairly close between the old and new tests. The new tests pick up a few arcs that weren't covered before. However, there are some regressions in code coverage. These are the ones I spotted:

https://github.com/ansible/ansible/blob/d18d4f84ecb28547220642b39bde08fd47615f0d/lib/ansible/plugins/action/reboot.py#L127-L128

https://github.com/ansible/ansible/blob/d18d4f84ecb28547220642b39bde08fd47615f0d/lib/ansible/plugins/action/reboot.py#L334-L337

webknjaz committed:
Drop redundant _low_level_execute_command double patch

This seems to duplicate the previously used `monkeypatch.setattr()` and shouldn't change anything.

Co-authored-by: Matt Clay <matt@mystile.com>

webknjaz committed:
Drop py2/3 compat boilerplate from `test_reboot`

Co-authored-by: Matt Clay <matt@mystile.com>

webknjaz committed:
Transform if-blocks into a mapping lookup @ `test_reboot`

Co-authored-by: Matt Clay <matt@mystile.com>

webknjaz committed:
Extend test_reboot_command w/ spaceless reboot cmd

This extends test coverage for the case when a no-arg command is
provided that the internal machinery returns an empty string for the
reboot command CLI options.

webknjaz committed:
Test connection drop post reboot

This covers an except block code path that happens when a reboot
command caused a connection reset before said connection had a
chance to shut down gracefully.

ci_complete
ci_coverage

webknjaz commented:
@mattclay I think I managed to cover those spots you pointed out..

mattclay commented:
@webknjaz Could you update this PR so the last commit includes `ci_complete` and `ci_coverage` in the comment? That should allow us to more easily make a comparison between code coverage from `devel` and this PR.

webknjaz commented:
done

mattclay reviewed (approved):
Nice. It looks like we're getting code coverage for everything we were previously, and more.

webknjaz commented:
Backport PRs:
- https://github.com/ansible/ansible/pull/81266
- https://github.com/ansible/ansible/pull/81267
- https://github.com/ansible/ansible/pull/81268
